Inventory.md

```diff
@@ -0,0 +1,173 @@
+
+### Addresses
+
+We can trivially re-encode addresses to work with the wallet core and re-decode them later
+(when checking if they are ours or not).
+
+We might be able to treat addresses more abstractly and avoid this.
+
+### Active slot coeff
+
+While missing in Byron, it will exist in Shelley. We could set it to 100% internally in the wallet. Should be a problem at all. ✅
+
+### Superfluous prevBlockHash in BlockHeader
+
+With a good abstraction, or dropped Jörmungandr support we could remove the prevBlockHash field in BlockHeader, but it does no damage there. ✅
+
+### Point ByronBlock vs BlockHeader
+
+We can probably get by with magic values to represent the `Origin`-case. ✅
+
+A way to treat points (like ourobouros-network's`Point ByronBlock`) more abstract in the wallet could potentially be needed to avoid sabotaging other abstractions.
+
+### SlotId vs SlotNo
+
+When slotsPerEpoch cannot change the conversion between SlotId and SlotNo is simple.
+
+But to support protocol parameter updates, we should consider making the core wallet logic deal with SlotNo for the Haskell node.
+
+For Byron, we can assume they are static. ✅
+
+### Epoch Bounday Blocks
+
+I imagine having two blocks with the same slot might break assumptions in our DB-rollbacks, but I don't know.
+
+A simple solution would be to filter them out. ✅
+
+For mainnet Shelley, EBBs do provide an easy way of knowing when a new epoch starts without having to keep track of protocol parameters (and when slots are counted from genesis). This would be an argument for not filtering them out. We will however most likely be maintaining up-to-date protocol parameters anyway, and would be able to tell when we're entering a new epoch through the protocol-params and the current slotNo.
+
+### No rollbacks further than "k"
+
+With the Haskell node we will be able to say that old enough state/checkpoints are immutable and can never be rolled back. We might want to change which checkpoints we store in DB. There is no point in keeping multiple checkpoints in the immutable part of the chain.
+
+### Delegation features
+
+We cannot implement delegation features. We could
+1. Keep the API the same, and use dummy implementations internally. E.g return an empty list of stake pools. E.g say that wallets have 0 account balance. I think this should work without any hustle.
+2. Make the wallet modular enough to not have to do the above. I am not sure how difficult this would be.
+
+## Other, fuzzier difficulties of integration
+
+I tried to get a wallet executable working with the Haskell node and discovered new kinds of difficulties.
+
+- Chain parameters are not in the Haskell genesis block. The wallet will need to hard-code or read them from a config file at startup.
+- The genesis block does not have to be treated differently because of above. The jörmungandr wallet currently treats it completely separate.
```

Comment:
Hmmm. No it doesn't treat the genesis block differently apart from using it as a starting point when instantiating a wallet. But in essence, it could be any block.

Reply:
>  apart from using it as a starting point when instantiating a wallet

Exactly. If we didn't have to support jörmungandr, the easiest thing would probably be to let wallet's be created completely empty, and apply the genesis block (without special treatment) when we start the restoration.

Reply:
No. No. No. Being unable to create empty wallet is a feature BY DESIGN. Because "an empty wallet" with no blocks requires handling the initialization as a special case, which means more logic, and increased chances of introducing bugs. This has nothing to do with Jörmungandr.

Reply:
>  Because "an empty wallet" with no blocks requires handling the initialization as a special case

Could you elaborate on what this refers to?

In my quick prototype I changed `initWallet` to take a block header instead of a full block, and I didn't see any problems.

https://github.com/input-output-hk/cardano-wallet/blob/ed10bfeabac2e184a3d090d875846091249a3288/lib/core/src/Cardano/Wallet/Primitive/Model.hs#L187-L197

Reply:
What happened to the transactions in the genesis block :) ? 
So now, we have a wallet that has already applied the given block header, but it has missed all the transactions from the genesis block. How do we make sure that we still get to apply the corresponding block, while still making sure that we don't apply the same block twice ?

Reply:
I assume that is why we should initialise our wallets with `Origin :: Point ByronBlock`!

(Haven't actually confirmed this, but I assume `Origin` is the point before genesis.

This is not in my prototype, you're right it's probably not syncing the genesis block.)

Reply:
> the chain-sync yield the genesis block.

I hardly believe that. The genesis block weights a hundred of megabytes. I don't see this being transmitted over the network just like this, but I may be wrong.

Reply:
Actually, you might be right 🤔 👀

Reply:
Seems like you're right 👍  https://github.com/input-output-hk/cardano-ledger/blob/0ad59ce46b3da8af6f6af97064e7c77cc7c2ac8b/cardano-ledger/src/Cardano/Chain/UTxO/GenesisUTxO.hs#L19-L36

Although I think it might technically not be a "block".

I also tried writing the first few blocks from chain-sync to disk, starting at `Origin`, and they  were all ≈670 bytes.